genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -2914,9 +2973,839 @@ def support_driver(self, direction, i_g, i_b, i_o, shrink_sphere):
         elif geom_type == gs.GEOM_TYPE.TERRAIN:
             if ti.static(self._solver.collider._has_terrain):
                 v, vid = self.support_field._func_support_prism(direction, i_g, i_b)
-        elif geom_type == gs.GEOM_TYPE.MESH and self._solver._enable_mujoco_compatibility:
+        elif geom_type == gs.GEOM_TYPE.MESH and self._enable_mujoco_compatibility:
             # If mujoco-compatible, do exhaustive search for the vertex
             v, vid = self.support_mesh(direction, i_g, i_b, i_o)
         else:
             v, vid = self.support_field._func_support_world(direction, i_g, i_b)
         return v, vid
+
+    @ti.func
+    def func_safe_gjk(self, i_ga, i_gb, i_b):
+        """
+        Safe GJK algorithm to compute the minimum distance between two convex objects.
+
+        This implementation is safer than the one based on the MuJoCo implementation for the following reasons:
+        1) It guarantees that the origin is strictly inside the tetrahedron when the intersection is detected.
+        2) It guarantees to generate a non-degenerate tetrahedron if there is no numerical error, which is necessary
+        for the following EPA algorithm to work correctly.
+        3) When computing the face normals on the simplex, it uses a more robust method than using the origin.
+
+        TODO: This implementation could be improved by using shrink_sphere option as the MuJoCo implementation does.
+        TODO: This implementation could be further improved by referencing the follow-up work shown below.
+
+        .. seealso::
+        Original paper:
+        Gilbert, Elmer G., Daniel W. Johnson, and S. Sathiya Keerthi.
+        "A fast procedure for computing the distance between complex objects in three-dimensional space."
+        IEEE Journal on Robotics and Automation 4.2 (2002): 193-203.
+
+        Further improvements:
+        Cameron, Stephen. "Enhancing GJK: Computing minimum and penetration distances between convex polyhedra."
+        Proceedings of international conference on robotics and automation. Vol. 4. IEEE, 1997.
+        https://www.cs.ox.ac.uk/people/stephen.cameron/distances/gjk2.4/
+
+        Montaut, Louis, et al. "Collision detection accelerated: An optimization perspective."
+        https://arxiv.org/abs/2205.09663
+        """
+        dir0 = gs.ti_vec3(0, 0, 1)
+        dir1 = gs.ti_vec3(0, 1, 0)
+
+        # Compute the initial tetrahedron using two random directions
+        init_flag = RETURN_CODE.SUCCESS
+        self.simplex[i_b].nverts = 0
+        for i in range(4):
+            dir = gs.ti_vec3(0.0, 0.0, 0.0)
```

Comment:
Not a huge fun of this `dir0`, `dir1` that are only used once, along with some complex logic. In my view, it obfuscates what you are trying to achieve. Why not just:
```python
dir = ti.Vector.zero(gs.ti_float, 3)
dir[i // 2] = 1.0 if i % 2 == 0 else -1.0
```
Or even:
```python
dir = ti.Vector.zero(gs.ti_float, 3)
dir[i // 2] = 1.0 - 2.0 * (i % 2)
```